### PR TITLE
XML optLong/getLong equivalent updates for string to number conversion.

### PR DESCRIPTION
**Pull request by rudrajyotib**
[stleary update]: See #790
Moved the code logic to a common utility to de-duplicate.

### CODE REVIEW

File: src/test/java/org/json/junit/JSONMLTest.java [hunk 1]
```diff
@@ -709,7 +709,7 @@ public void commentsInXML() {
     @Test
     public void testToJSONArray_jsonOutput() {
         final String originalXml = "<root><id>01</id><id>1</id><id>00</id><id>0</id><item id=\"01\"/><title>True</title></root>";
-        final String expectedJsonString = "[\"root\",[\"id\",\"01\"],[\"id\",1],[\"id\",\"00\"],[\"id\",0],[\"item\",{\"id\":\"01\"}],[\"title\",true]]";
+        final String expectedJsonString = "[\"root\",[\"id\",1],[\"id\",1],[\"id\",0],[\"id\",0],[\"item\",{\"id\":1}],[\"title\",true]]";
```
Comment (stleary) [inline src/test/java/org/json/junit/JSONMLTest.java:712]: Please do not remove any test cases. Adding new cases is fine.

File: src/test/java/org/json/junit/XMLConfigurationTest.java [hunk 2]
```diff
@@ -733,7 +733,7 @@ public void contentOperations() {
     @Test
     public void testToJSONArray_jsonOutput() {
         final String originalXml = "<root><id>01</id><id>1</id><id>00</id><id>0</id><item id=\"01\"/><title>True</title></root>";
-        final JSONObject expected = new JSONObject("{\"root\":{\"item\":{\"id\":\"01\"},\"id\":[\"01\",1,\"00\",0],\"title\":true}}");
+        final JSONObject expected = new JSONObject("{\"root\":{\"item\":{\"id\":1},\"id\":[1,1,0,0],\"title\":true}}");
```
Comment (stleary) [inline src/test/java/org/json/junit/XMLConfigurationTest.java:736]: Please do not remove existing test cases.

File: src/test/java/org/json/junit/XMLTest.java [hunk 3]
```diff
@@ -791,7 +791,7 @@ private void compareFileToJSONObject(String xmlStr, String expectedStr) {
     @Test
     public void testToJSONArray_jsonOutput() {
         final String originalXml = "<root><id>01</id><id>1</id><id>00</id><id>0</id><item id=\"01\"/><title>True</title></root>";
-        final JSONObject expectedJson = new JSONObject("{\"root\":{\"item\":{\"id\":\"01\"},\"id\":[\"01\",1,\"00\",0],\"title\":true}}");
+        final JSONObject expectedJson = new JSONObject("{\"root\":{\"item\":{\"id\":1},\"id\":[1,1,0,0],\"title\":true}}");
```
Comment (stleary) [inline src/test/java/org/json/junit/XMLTest.java:794]: Please do not remove existing test cases.

**Comment by stleary** [comment]
It would be better to open an issue first, which would provide a forum to discuss the changes before submitting a PR. 
It's not a bad idea, but you will need to provide evidence that no existing functionality has changed in JSONArray, JSONObject, or XML. Just passing the unit tests is not sufficient.

File: src/test/java/org/json/junit/JSONMLTest.java (hunk 1 above)
Comment (rudrajyotib) [reply src/test/java/org/json/junit/JSONMLTest.java:712]: Hello @stleary  -These test cases were misleading. In the expected JSON string, the numbers still had preceding zeros. If we want the values to be treated as numbers, the leading zeros should go. In fact, the test comment also confirms that.

File: src/test/java/org/json/junit/XMLConfigurationTest.java (hunk 2 above)
Comment (rudrajyotib) [reply src/test/java/org/json/junit/XMLConfigurationTest.java:736]: Hello @stleary  - hese test cases were misleading. In the expected JSON string, the numbers still had preceding zeros. If we want the values to be treated as numbers, the leading zeros should go. In fact, the test comment also confirms that.

File: src/test/java/org/json/junit/XMLTest.java (hunk 3 above)
Comment (rudrajyotib) [reply src/test/java/org/json/junit/XMLTest.java:794]: Hello @stleary - hese test cases were misleading. In the expected JSON string, the numbers still had preceding zeros. If we want the values to be treated as numbers, the leading zeros should go. In fact, the test comment also confirms that.

**Comment by stleary** [comment]
Closed due to failing existing unit tests.

**Comment by stleary** [comment]
@rudrajyotib Sorry about that, should not have closed this PR. It still needs work. For example, if a unit test must be broken, call it out and provide a convincing explanation about why it was necessary. Also, I would strongly discourage refactoring large amounts of code. It's better to have duplicate code at first, and then refactor it in a later PR.

**Comment by rudrajyotib** [comment]
@stleary - can you please add the Hacktoberfest tag back. I have reverted all changes that combined feature and refactor. We can now apply the changes one by one.

**Comment by rudrajyotib** [comment]
@stleary - new implementation breaks some existing test cases.
Expected :["root",["id","01"],["id",1],["id","00"],["id",0],["item",{"id":"01"}],["title",true]]
Actual   :["root",["id",1],["id",1],["id",0],["id",0],["item",{"id":1}],["title",true]]

This test case being broken is the expected behavior, in my opinion. Please suggest, if I shall commit with the test cases expectation changed.

**Comment by johnjaylward** [comment]
@rudrajyotib , There has been a lot of PRs merged to master today. Please do a merge or rebase to get the latest changes.

**Comment by rudrajyotib** [comment]
@stleary - now synced and merged with master.

**Comment by stleary** [comment]
**What problem does this code solve?**
The changes that were made for optLong() vs getLong() should also be made to XML

**Risks**
Low

**Changes to the API?**
Affects how XML converts strings to numbers

**Will this require a new release?**
No

**Should the documentation be updated?**
No

**Does it break the unit tests?**
The unit tests were updated for the new behavior

**Was any code refactored in this commit?**
No

**Review status**
**APPROVED**

Starting 3-day comment window

**Comment by stleary** [comment]
@rudrajyotib I will update your PRs, but I don't think it's necessary since all accepted PRs get an approving review. I only keep the hacktoberfest label for internal tracking. From the website:
"YOUR PR/MRS MUST BE MERGED, HAVE THE “HACKTOBERFEST-ACCEPTED” LABEL, OR HAVE AN OVERALL APPROVING REVIEW."

**Pull request merged**